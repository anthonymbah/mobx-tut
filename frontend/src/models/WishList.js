import { types } from 'mobx-state-tree';


const WishListItem = types.model({
	name: types.string,
	price: types.number,
}).actions(self => ({
	changeName(name){
		self.name = name;
	},
	changePrice(price){
		self.price = price;
	}
}));

const WishList = types.model({
	items: types.optional(types.array(WishListItem), []),
}).actions(self => ({
	add(item){
		self.items.push(item);
	}
})).views(self => ({
	get totalPrice(){
		return self.items.reduce((acc,item) => acc + item.price, 0);
	}
}));

export {
	WishList,
	WishListItem,
}