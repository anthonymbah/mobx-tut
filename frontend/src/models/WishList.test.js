import { getSnapshot, onSnapshot, onPatch } from 'mobx-state-tree';
import { reaction } from 'mobx';
import { WishList, WishListItem } from './WishList';

it('can create an instance of the model', () => {

	const item = WishListItem.create({
		name: 'book One',
		price: 23.00
	});

	expect(getSnapshot(item)).toEqual({
			name: 'book One',
			price: 23.00,
	});
});

it('can create an instance of list of model', () => {

	const list = WishList.create()

	list.add({
		name: 'book One',
		price: 23.00
	});

	expect(getSnapshot(list)).toEqual({
		items:[{
			name: 'book One',
			price: 23.00,
		}]
	});

});

it('can create an instance of list of model: snapshot', () => {

	const list = WishList.create()

	list.add({
		name: 'book One',
		price: 23.00
	});

	expect(getSnapshot(list)).toMatchSnapshot();
});

it('can create an instance of list of model on state change', () => {

	const list = WishList.create();
	const states = [];
	onSnapshot(list, snapshot => {
		states.push(snapshot);
	});

	list.add({
		name: 'book One',
		price: 23.00
	});

	list.items[0].changeName('book X');

	expect(states).toMatchSnapshot();
});

it('can create an instance of list of model on patch change', () => {

	const list = WishList.create();
	const patches = [];
	onPatch(list, patch => {
		patches.push(patch);
	});

	list.add({
		name: 'book One',
		price: 23.00
	});

	list.items[0].changeName('book X');

	expect(patches).toMatchSnapshot();
});

it('can total price', () => {

	const list = WishList.create();
	list.add({
		name: 'book One',
		price: 23.00
	});
	let changed = 0;
	reaction(() => list.totalPrice, () => changed ++);
	
	const [ item ] = list.items;
	item.changeName('book X');
	expect(changed).toBe(0);

	item.changePrice(24);
	expect(changed).toBe(1);

});