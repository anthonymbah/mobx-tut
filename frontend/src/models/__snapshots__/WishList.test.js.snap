// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create an instance of list of model on patch change 1`] = `
Array [
  Object {
    "op": "add",
    "path": "/items/0",
    "value": Object {
      "name": "book One",
      "price": 23,
    },
  },
  Object {
    "op": "replace",
    "path": "/items/0/name",
    "value": "book X",
  },
]
`;

exports[`can create an instance of list of model on state change 1`] = `
Array [
  Object {
    "items": Array [
      Object {
        "name": "book One",
        "price": 23,
      },
    ],
  },
  Object {
    "items": Array [
      Object {
        "name": "book X",
        "price": 23,
      },
    ],
  },
]
`;

exports[`can create an instance of list of model: snapshot 1`] = `
Object {
  "items": Array [
    Object {
      "name": "book One",
      "price": 23,
    },
  ],
}
`;
